source/opt/trim_capabilities_pass.cpp

```diff
@@ -0,0 +1,197 @@
+//
+// Handlers names follow the following convention:
+//  Handler_<Opcode>_<Capability>()
+
+static std::optional<spv::Capability> Handler_OpReadClockKHR_ShaderClockKHR(
+    const IRContext*, const Instruction*) {
+  // The simple presence of this opcode is enough to determine the capability
+  // requirement.
+  return spv::Capability::ShaderClockKHR;
+}
+
+// Opcode of interest to determine capabilities requirements.
+static const std::unordered_multimap<spv::Op, OpcodeHandler, UnsignedHash>
+    kOpcodeHandlers{
+        {spv::Op::OpReadClockKHR, Handler_OpReadClockKHR_ShaderClockKHR},
+    };
+
+// ==============  End opcode handler implementations.  =======================
+
+CapabilitySet TrimCapabilitiesPass::DetermineRequiredCapabilities() {
+  CapabilitySet required_capabilities;
+  const IRContext* ctx = context();
+
+  get_module()->ForEachInst(
+      [ctx, &required_capabilities](Instruction* instruction) {
+        auto[begin, end] = kOpcodeHandlers.equal_range(instruction->opcode());
+        std::for_each(begin, end,
+                      [&required_capabilities, ctx, instruction](auto& item) {
+                        auto result = item.second(ctx, instruction);
+                        if (result.has_value()) {
+                          required_capabilities.insert(*result);
+                        }
+                      });
+      });
+
+  return required_capabilities;
+}
+
+// Utility function to, given a list of capabilities, retrieve the associated
+// extensions.
+template <typename Container>
+static ExtensionSet GetExtensionsForCapabilities(
+    const Container& capabilities) {
+  ExtensionSet output;
+  for (auto capability : capabilities) {
+    auto extension = kCapabilityExtensionMapping.find(capability);
+    if (extension != kCapabilityExtensionMapping.end()) {
+      output.insert(extension->second);
+    }
+  }
+  return output;
+}
+
+void TrimCapabilitiesPass::TrimExtensionsForCapabilities(
+    const CapabilitySet& required_capabilities,
+    const std::vector<spv::Capability>& trimmed_capabilities) {
+  const ExtensionSet required_extensions =
+      GetExtensionsForCapabilities(required_capabilities);
+  const ExtensionSet trimmed_extensions =
+      GetExtensionsForCapabilities(trimmed_capabilities);
+
+  std::vector<Instruction*> instructions_to_kill;
+  for (auto& instruction : context()->extensions()) {
+    assert(instruction.NumOperands() == 1);
+    auto extension = instruction.GetOperand(0).AsString();
+
+    if (required_extensions.count(extension) != 0) {
+      continue;
+    }
+
+    if (trimmed_extensions.count(extension) != 0) {
+      instructions_to_kill.push_back(&instruction);
+    }
+  }
+
+  // Killing an instruction invalidates the instruction iterator. We must gather
+  // & kill in 2 steps.
+  for (auto instruction : instructions_to_kill) {
+    context()->KillInst(instruction);
+  }
+}
+
+Pass::Status TrimCapabilitiesPass::FilterModuleCapabilities(
```

Comment:
This remove the extensions as well as the capabilities. It might be good it have three functions instead of two:

```
// One function to remove the OpCapability instructions, and return the capabilities that were removed.
CapabilitySet TrimUnrequiredCapabilities(const CapabilitySet& required_capabilities);

// The same as the existing function.
void TrimExtensionsForCapabilities(required_capabilities, capabilities_to_trim);

// 
Pass::Status TrimCapabilitiesPass::TrimUnrequiredCapabilitiesAndExtensions(const CapabilitySet& required_capabilities) {
  CapabilitySet trimmed_capabilities =  TrimUnrequiredCapabilities(required_capabilities);
  TrimExtensionsForCapabilities(required_capabilities, trimmed_capabilities);
}
```

Reply:
valid point. Removed the filter function (inlined in the pass function), and added TrimExtensionsForCapabilities.